src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -1393,14 +1403,30 @@ MulticopterPositionControl::task_main()
 
 				/* use constant descend rate when landing, ignore altitude setpoint */
 				if (!_control_mode.flag_control_manual_enabled && _pos_sp_triplet.current.valid
-						&& _pos_sp_triplet.current.type == position_setpoint_s::SETPOINT_TYPE_LAND) {
+				    && _pos_sp_triplet.current.type == position_setpoint_s::SETPOINT_TYPE_LAND) {
 					_vel_sp(2) = _params.land_speed;
+
+					/* final approach, decrease descent speed */
+					float _dist_ground = (_local_pos.dist_bottom_valid) ? _local_pos.dist_bottom : -_local_pos.z;
+
+					if (_local_pos.dist_bottom_valid || _local_pos.z_valid) {
+						if (_params.land_final_approach_alt > 0.0f && _dist_ground <= _params.land_final_approach_alt + 1.0f) {
+							/* decrease decent velocity 1m above fna_alt */
+							if (_dist_ground > _params.land_final_approach_alt) {
+								_vel_sp(2) = _params.land_final_approach_speed + ((_dist_ground - _params.land_final_approach_alt) *
```

Comment:
This deceleration shouldn't be necessary because vertical acceleration is handled below.